From: Guilhem Moulin <guilhem@debian.org>
Date: Fri, 15 Jul 2022 19:40:53 +0200
Subject: unit-wipe-test: Skip DIO tests when the file system doesn't support
 O_DIRECT.

Otherwise the test fails with -EINVAL when the source tree (and thus
$FILE) is on a file system such as tmpfs which doesn't support O_DIRECT.
This appears to be the case on the Debian build daemons and causes
cryptsetup 2:2.5.0~rc1-1 to FTBFS.

Forwarded: https://gitlab.com/cryptsetup/cryptsetup/-/merge_requests/387
---
 tests/unit-wipe-test | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/tests/unit-wipe-test b/tests/unit-wipe-test
index 208e05f..d6c99b5 100755
--- a/tests/unit-wipe-test
+++ b/tests/unit-wipe-test
@@ -119,8 +119,10 @@ add_file
 
 echo -n "[1] Wipe full file "
 for bs in 0 $MB_BYTES $((4*$MB_BYTES)); do
-	echo -n [$bs/DIO]
-	test_wipe_full $FILE $bs
+	if dd if=$FILE of=/dev/null iflag=direct >/dev/null 2>&1; then
+		echo -n [$bs/DIO]
+		test_wipe_full $FILE $bs
+	fi
 	echo -n [$bs]
 	test_wipe_full $FILE $bs no-dio
 done
@@ -128,8 +130,10 @@ echo "[OK]"
 
 echo -n "[2] Wipe blocks in file "
 for bs in 0 $MB_BYTES $((4*$MB_BYTES)); do
-	echo -n [$bs/DIO]
-	test_wipe_blocks $FILE $bs
+	if dd if=$FILE of=/dev/null iflag=direct >/dev/null 2>&1; then
+		echo -n [$bs/DIO]
+		test_wipe_blocks $FILE $bs
+	fi
 	echo -n [$bs]
 	test_wipe_blocks $FILE $bs no-dio
 done
